### ISSUE TITLE
google.api_core.exceptions.ServiceUnavailable: 503 failed to connect to all addresses

### ISSUE THREAD

**Issue by radianceltd** [GoogleCloudPlatform/python-docs-samples#2638]
Every time I make a device request, there will be problems as long as I get to this code

```
python=3.7.0
cryptography==2.7
flaky==3.6.1
gcp-devrel-py-tools==0.0.15
google-api-python-client==1.7.11
google-auth-httplib2==0.0.3
google-auth==1.6.3
google-cloud-iot==0.3.0
google-cloud-pubsub==1.0.0
oauth2client==4.1.3
paho-mqtt==1.4.0
pyjwt==1.7.1
```

Error code

```
        client = iot_v1.DeviceManagerClient()

        parent = client.registry_path(project_id, cloud_region, registry_id)

        device_path = client.device_path(
            'tmw025', 'us-central1', 'my-registry', 'device-002')

        device = client.get_device(device_path)  client = iot_v1.DeviceManagerClient()

        parent = client.registry_path(project_id, cloud_region, registry_id)

        device_path = client.device_path(
            'tmw025', 'us-central1', 'my-registry', 'device-002')
# Error is reported when the code is executed here
        device = client.get_device(device_path)

```
I'm going to have a problem executing this code
```
device = client.get_device(device_path)
```
I have projects, devices and region in my Google cloud platform, But I can't get the data

```

Traceback (most recent call last):
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/grpc/_channel.py", line 690, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/grpc/_channel.py", line 592, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses"
	debug_error_string = "{"created":"@1576717282.105640000","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":3934,"referenced_errors":[{"created":"@1576717282.105638000","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":393,"grpc_status":14}]}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/retry.py", line 184, in retry_target
    return target()
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/timeout.py", line 214, in func_with_timeout
    return func(*args, **kwargs)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.ServiceUnavailable: 503 failed to connect to all addresses

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/rnd/Desktop/AllProject/Python/iotcore_test/app/web/googlecore.py", line 37, in device
    algorithm)
  File "/Users/rnd/Desktop/AllProject/Python/iotcore_test/app/spider/google_core.py", line 82, in create_gateway
    device = client.get_device(device_path)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/cloud/iot_v1/gapic/device_manager_client.py", line 792, in get_device
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func
    on_error=on_error,
  File "/Users/rnd/.local/share/virtualenvs/iotcore_test-cNntKDwd/lib/python3.7/site-packages/google/api_core/retry.py", line 206, in retry_target
    last_exc,
  File "<string>", line 3, in raise_from
google.api_core.exceptions.RetryError: Deadline of 120.0s exceeded while calling functools.partial(<function _wrap_unary_errors.<locals>.error_remapped_callable at 0x10cdd2950>, name: "projects/tmw025/locations/us-central1/registries/my-registry/devices/device-002"
, metadata=[('x-goog-request-params', 'name=projects/tmw025/locations/us-central1/registries/my-registry/devices/device-002'), ('x-goog-api-client', 'gl-python/3.7.0 grpc/1.25.0 gax/1.15.0 gapic/0.3.0')]), last exception: 503 failed to connect to all addresses

```

**Comment by radianceltd** [comment]
Whether login authorization verification is required？

**Comment by gguuss** [comment]
This might be a duplicate of [this issue](https://github.com/GoogleCloudPlatform/python-docs-samples/issues/2626).

**Comment by gguuss** [comment]
Here is configuration I have confirmed to work:

Python 3.6 / 
OS: Mac OSX

Output from pip freeze:

```
asn1crypto==1.2.0
astroid==2.3.3
cachetools==4.0.0
certifi==2019.11.28
cffi==1.13.2
chardet==3.0.4
cryptography==2.7
flaky==3.6.1
gcp-devrel-py-tools==0.0.15
google-api-core==1.15.0
google-api-python-client==1.7.11
google-auth==1.6.3
google-auth-httplib2==0.0.3
google-cloud-iot==0.3.0
google-cloud-pubsub==1.0.0
googleapis-common-protos==1.6.0
grpc-google-iam-v1==0.12.3
grpcio==1.26.0
httplib2==0.15.0
idna==2.8
isort==4.3.21
lazy-object-proxy==1.4.3
mccabe==0.6.1
oauth2client==4.1.3
packaging==19.2
paho-mqtt==1.4.0
protobuf==3.11.1
pyasn1==0.4.8
pyasn1-modules==0.2.7
pycparser==2.19
PyJWT==1.7.1
pylint==2.4.4
pyparsing==2.4.5
pytz==2019.3
requests==2.22.0
retrying==1.3.3
rsa==4.0
six==1.13.0
typed-ast==1.4.0
uritemplate==3.0.0
urllib3==1.25.7
wrapt==1.11.2
```

From your reports, it's difficult to separate your flask errors from the Google errors.

Can you try the following:

Clone the Google Cloud Python repo:

    git clone https://github.com/googlecloudplatform/python-docs-samples

Navigate to the IoT Core sample:

    cd python-docs-samples/iot/api-client/manager

Create a virtual environment:

    virtualenv -p python3 env3

Start your virtual environment:

    source env3/bin/activate
    pip install -r requirements.txt

Setup the sample to run:

    export GOOGLE_APPLICATION_CREDENTIALS=<your-path-to-service.json>
    export GOOGLE_CLOUD_PROJECT=<your-project-id>
    export GCLOUD_PROJECT=$GOOGLE_CLOUD_PROJECT

Try listing the registries in your project:

    python manager.py list

**Comment by radianceltd** [comment]
```
(env3) RNDdeiMac-2:manager rnd$ pip list
Package                  Version   
------------------------ ----------
asn1crypto               1.2.0     
astroid                  2.3.3     
cachetools               4.0.0     
certifi                  2019.11.28
cffi                     1.13.2    
chardet                  3.0.4     
cryptography             2.7       
flaky                    3.6.1     
gcp-devrel-py-tools      0.0.15    
google-api-core          1.15.0    
google-api-python-client 1.7.11    
google-auth              1.6.3     
google-auth-httplib2     0.0.3     
google-cloud-iot         0.3.0     
google-cloud-pubsub      1.0.0     
googleapis-common-protos 1.6.0     
grpc-google-iam-v1       0.12.3    
grpcio                   1.26.0    
httplib2                 0.15.0    
idna                     2.8       
isort                    4.3.21    
lazy-object-proxy        1.4.3     
mccabe                   0.6.1     
oauth2client             4.1.3     
packaging                19.2      
paho-mqtt                1.4.0     
pip                      19.3.1    
protobuf                 3.11.1    
pyasn1                   0.4.8     
pyasn1-modules           0.2.7     
pycparser                2.19      
PyJWT                    1.7.1     
pylint                   2.4.4     
pyparsing                2.4.5     
pytz                     2019.3    
requests                 2.22.0    
retrying                 1.3.3     
rsa                      4.0       
setuptools               42.0.2    
six                      1.13.0    
typed-ast                1.4.0     
uritemplate              3.0.1     
urllib3                  1.25.7    
wheel                    0.33.6    
wrapt                    1.11.2    
(env3) RNDdeiMac-2:manager rnd$ 


```
(env3) RNDdeiMac-2:manager rnd$ deactivate
RNDdeiMac-2:manager rnd$ python manager.py list-registries
Traceback (most recent call last):
  File "manager.py", line 37, in <module>
    from google.api_core.exceptions import AlreadyExists
ImportError: No module named google.api_core.exceptions
RNDdeiMac-2:manager rnd$ 
Why does this happen？

**Comment by radianceltd** [comment]
How to set up a firewall to unblock grpc call traffic？

**Comment by radianceltd** [comment]
I think it's a firewall problem, but I don't know how to set it up

**Comment by gguuss** [comment]
To check whether it's a firewall problem you may want to try using the [Cloud Shell](https://cloud.google.com/shell/) to test your code first to make sure it's not the code. Issues from a firewall would be on-device or on-network blocking of specific ports. I'm not an expert in re-routing via VPN, proxying, or tunneling the GRPC communication but those would be your options to work around something blocking traffic.

**Comment by radianceltd** [comment]
But I ran it on the gccloud command line and found no problem
```
gcloud iot registries create REGISTRY_ID \
    --project=PROJECT_ID \
    --region=REGION \
    [--event-notification-config=topic=TOPIC,[subfolder=SUBFOLDER] [--event-notification-config=...]]
    [--state-pubsub-topic=STATE_PUBSUB_TOPIC]

```

**Comment by radianceltd** [comment]
I have no problem with cloud shell testing
```
radiancegooplay@cloudshell:~/python-docs-samples/iot/api-client/manager (tmw025)$ source env3/bin/activate
(env3) radiancegooplay@cloudshell:~/python-docs-samples/iot/api-client/manager (tmw025)$ python manager.py list-registries
Listing Registries
id: my-registry
        name: projects/tmw025/locations/us-central1/registries/my-registry
id: mytest-registry
        name: projects/tmw025/locations/us-central1/registries/mytest-registry
(env3) radiancegooplay@cloudshell:~/python-docs-samples/iot/api-client/manager (tmw025)$
```

**Comment by gguuss** [comment]
I think the next step for you is going to be figuring out why the port is being blocked when you run on your target hardware. Try and figure out if you have a port blocker / firewall that is causing your issue.

**Comment by radianceltd** [comment]
Okay, I'll try. Tank u

**Comment by gguuss** [comment]
@radianceltd I'm going to close this issue, please reopen if you're still encountering issues with the sample.

**Comment by codal-mharkhani** [comment]
@radianceltd @gguuss 
is this issue resolved...? if yes can you please describe how.
